daphne_worker/src/durable/mod.rs

```diff
@@ -2,305 +2,169 @@
-
-    fn env(&self) -> &Env;
-
-    /// Run garbage collection requests.
-    ///
-    /// If a garbage collection request is handled no further processing needs to be done, as such,
-    /// this function will consume the request, otherwise it will return the passed in request for
-    /// further handling.
-    async fn schedule_for_garbage_collection(
-        &mut self,
-        req: Request,
-    ) -> Result<std::ops::ControlFlow<(), Request>> {
-        match GarbageCollector::try_from_uri(&req.path()) {
-            Some(GarbageCollector::DeleteAll) => {
-                self.state().storage().delete_all().await?;
-                *self.touched() = false;
-                return Ok(std::ops::ControlFlow::Break(()));
-            }
-            _ if !*self.touched() => {
-                // The GarbageCollector should only be used when running tests. In production, the DO->DO
-                // communication overhead adds unacceptable latency, and there's no need to do the
-                // bulk deletes of state that test suites require.
-                if matches!(self.deployment(), DaphneWorkerDeployment::Dev) {
-                    let touched = state_set_if_not_exists(self.state(), "touched", &true)
-                        .await?
-                        .unwrap_or(false);
-                    if !touched {
-                        let durable = crate::durable::DurableConnector::new(self.env());
-                        durable
-                            .post(
-                                bindings::GarbageCollector::BINDING,
-                                bindings::GarbageCollector::Put.to_uri(),
-                                bindings::GarbageCollector::name(()).unwrap_from_name(),
-                                &crate::durable::DurableReference {
-                                    binding: Self::DurableMethod::BINDING.to_string(),
-                                    id_hex: self.state().id().to_string(),
-                                    task_id: None,
-                                },
-                            )
-                            .await?;
-                    }
-                }
-                *self.touched() = true;
-            }
-            _ => {}
+/// Ensure that an alarm for self deletion is setup for a durable object.
+async fn ensure_alarmed(state: &State, env: &Env, lifetime: ScheduledTime) -> Result<()> {
+    if let Err(e) = state.storage().set_alarm(lifetime).await {
+        if matches!(deployment(env), DaphneWorkerDeployment::Dev) {
+            warn!("ignoring get_alarm() failure in a dev environment until --experimental-local implements it: {e}");
+        } else {
+            // We only return an error if not in the "dev" deployment as
+            // the experimental-local dev environment doesn't have
+            // working get_alarm() and set_alarm() yet, so we want to
+            // ignore errors in that case.
+            return Err(e);
```

Comment:
Here we unconditionally set the alarm, whereas previously we only set the alarm we haven't already. This is intended to change GC semantics such that the DO instance is deleted some time after last touch, rather than on first touch as we did before. I think this is a great idea, but two points:
1. We need to document the change of semantics in the commit message. This is an important functional change.
2. Are there side-effects of calling `set_alarm()` multiple times? I'm wondering if (a) all calls after the first **are ignored** or (b) if calling `set_alarm()` on every DO request is inefficient. Is there any documentation that can help shed light?

Reply:
I tested this implementation and set_alarm does clear the previous alarm before creating a new one. On the topic of efficiency I have no idea tbh but I can't imagine it's that expensive, the docs don't mention anything.

docs: https://developers.cloudflare.com/durable-objects/api/alarms/

The docs don't explicitly mention this "replace last alarm" behavior but `getAlarm` seems to suggest it and experimentation confirms it. Speaking of which, running this locally didn't trigger the error that the code seems to suggest it would :thinking: I think this branch can be removed.

Reply:
We're using an old version of miniflare for the end-to-end tests, not the workers runtime itself. The behavior might be different in deployment

I'm not too worried about efficiency right now -- let's just deploy and monitor. Load testing should point out any issues.

I think the only concern I have left is ensuring that setting the alarm multiple times has the expected effect. If documentation supports this great. Otherwise, I'd propose testing it out before merging.